lib/signs_ui/config/sign_groups.ex

```diff
@@ -0,0 +1,50 @@
+defmodule SignsUi.Config.SignGroups do
+  @moduledoc """
+  Represents the state of all current sign groups.
+  """
+
+  alias SignsUi.Alerts.Alert
+  alias SignsUi.Config.SignGroup
+
+  @type route_id() :: String.t()
+  @type unix_time() :: String.t()
+  @type t() :: %{unix_time() => SignGroup.t() | %{}}
+  @type by_route() :: %{route_id() => t()}
+
+  @doc """
+  Returns a map of expired group timestamps to empty maps.
+  """
+  @spec expired(t(), DateTime.t(), MapSet.t(Alert.id())) :: t()
+  def expired(sign_groups, current_time, alerts) do
+    for {created_at, group} <- sign_groups,
+        SignGroup.expired?(group, current_time, alerts),
+        into: %{} do
+      {created_at, %{}}
+    end
+  end
+
+  @spec from_json(t()) :: t()
+  def from_json(groups) do
+    Map.new(groups, fn {time, group} -> {time, SignGroup.from_json(group)} end)
+  end
+
+  @spec by_route(t()) :: by_route()
+  def by_route(groups) do
+    groups
+    |> Enum.group_by(fn {_created_at, %SignGroup{route_id: route_id}} -> route_id end)
+    |> Enum.map(fn {route_id, groups} -> {route_id, Map.new(groups)} end)
+    |> Enum.into(%{})
+  end
+
+  @spec update({unix_time(), SignGroup.t() | %{}}, t()) :: t()
+  def update({created_at, map}, groups) when map == %{}, do: Map.delete(groups, created_at)
+
+  def update({created_at, %SignGroup{sign_ids: new_sign_ids} = group}, groups) do
+    groups
+    |> Enum.map(fn {created_at, %SignGroup{sign_ids: sign_ids} = group} ->
+      {created_at, %{group | sign_ids: Enum.filter(sign_ids, &(&1 not in new_sign_ids))}}
```

Comment:
Rather than `filter` / `not in`, maybe `Enum.reject` / `in` instead? Feels a little double-negativey.

Reply:
I always forget that `reject` is a thing!